pkg/storage/wal/manager.go

```diff
@@ -0,0 +1,213 @@
+// of being flushed. When the WAL is full, subsequent appends fail with ErrFull.
+// It is not permitted to append more data until another segment has been flushed
+// and returned to the available list. This allows the manager to apply back-pressure
+// and avoid congestion collapse due to excessive timeouts and retries.
+type Manager struct {
+	cfg Config
+
+	// available is a list of segments that are available and accepting data.
+	// All segments other than the segment at the front of the list are empty,
+	// and only the segment at the front of the list is written to. When this
+	// segment has exceeded its maximum age or maximum size it is moved to the
+	// pending list to be flushed, and the next segment in the available list
+	// takes its place.
+	available *list.List
+
+	// pending is a list of segments that are waiting to be flushed. Once
+	// flushed, the segment is reset and moved to the back of the available
+	// list to accept writes again.
+	pending  *list.List
+	shutdown chan struct{}
+	mu       sync.Mutex
+}
+
+// item is similar to PendingItem, but it is an internal struct used in the
+// available and pending lists. It contains a single-use result that is returned
+// to callers of Manager.Append() and a re-usable segment that is reset after
+// each flush.
+type item struct {
+	r *AppendResult
+	w *SegmentWriter
+	// firstAppendedAt is the time of the first append to the segment, and is
+	// used to know when the segment has exceeded the maximum age and should
+	// be moved to the pending list. It is reset after each flush.
+	firstAppendedAt time.Time
+}
+
+// PendingItem contains a result and the segment to be flushed. ClosedWriters
+// are to be returned following a flush so the segment can be re-used.
+type PendingItem struct {
+	Result *AppendResult
+	Writer *SegmentWriter
+}
+
+func NewManager(cfg Config) (*Manager, error) {
+	m := Manager{
+		cfg:       cfg,
+		available: list.New(),
+		pending:   list.New(),
+		shutdown:  make(chan struct{}),
+	}
+	for i := int64(0); i < cfg.MaxSegments; i++ {
+		w, err := NewWalSegmentWriter()
+		if err != nil {
+			return nil, err
+		}
+		m.available.PushBack(&item{
+			r: &AppendResult{done: make(chan struct{})},
+			w: w,
+		})
+	}
+	return &m, nil
+}
+
+func (m *Manager) Append(r AppendRequest) (*AppendResult, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.available.Len() == 0 {
+		return nil, ErrFull
+	}
+	el := m.available.Front()
+	it := el.Value.(*item)
+	if it.firstAppendedAt.IsZero() {
+		it.firstAppendedAt = time.Now()
+	}
+	it.w.Append(r.TenantID, r.LabelsStr, r.Labels, r.Entries)
+	// If the segment exceeded the maximum age or the maximum size, move it to
+	// the closed list to be flushed.
+	if time.Since(it.firstAppendedAt) >= m.cfg.MaxAge || it.w.InputSize() >= m.cfg.MaxSegmentSize {
+		m.pending.PushBack(it)
+		m.available.Remove(el)
+	}
+	return it.r, nil
+}
+
+func (m *Manager) Get() (*PendingItem, error) {
```

Comment:
mega nit: Maybe this should be called GetPending or something else?
API Clients only call Append but my mental model is that Get should return a valid WAL, not a closed one.
I don't feel strongly about it though.

Reply:
Yeah I don't have strong opinion on this either. I took the name `Get` and `Put` from `sync.Pool`.

Reply:
Let me think on this, because `PutPending` sounds weird and incorrect.